src/kernel.cr

```diff
@@ -123,24 +123,28 @@ end
 
 # :nodoc:
 module AtExitHandlers
-  @@running = false
-
   def self.add(handler)
     handlers = @@handlers ||= [] of Int32 ->
     handlers << handler
   end
 
+  @@handler_iterator : Iterator(Int32 ->)?
+
   def self.run(status)
-    return if @@running
-    @@running = true
+    return status unless handlers = @@handlers
 
-    @@handlers.try &.reverse_each do |handler|
+    handler_iterator = @@handler_iterator ||= handlers.reverse_each
```

Comment:
Actually, please write a note to explain we assign to `@@handler_iterator` to make sure we only call each handler once, even if an AtExitHandler calls `exit()`.

Reply:
Good idea, done!